extensions/standard-processors/processors/AttributesToJSON.cpp

```diff
@@ -0,0 +1,174 @@
+  core::PropertyBuilder::createProperty("Null Value")
+    ->withDescription("If true a non existing or empty attribute will be NULL in the resulting JSON. If false an empty string will be placed in the JSON.")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(false)
+    ->build());
+
+core::Relationship AttributesToJSON::Success("success", "All FlowFiles received are routed to success");
+
+void AttributesToJSON::initialize() {
+  setSupportedProperties({
+    AttributesList,
+    AttributesRegularExpression,
+    Destination,
+    IncludeCoreAttributes,
+    NullValue
+  });
+  setSupportedRelationships({Success});
+}
+
+void AttributesToJSON::onSchedule(core::ProcessContext* context, core::ProcessSessionFactory* /*sessionFactory*/) {
+  std::string value;
+  if (context->getProperty(AttributesList.getName(), value) && !value.empty()) {
+    attribute_list_ = utils::StringUtils::splitAndTrimRemovingEmpty(value, ",");
+  }
+  if (context->getProperty(AttributesRegularExpression.getName(), value) && !value.empty()) {
+    attributes_regular_expression_ = std::regex(value);
+  }
+  write_destination_ = WriteDestination::parse(utils::parsePropertyWithAllowableValuesOrThrow(*context, Destination.getName(), WriteDestination::values()).c_str());
+  context->getProperty(IncludeCoreAttributes.getName(), include_core_attributes_);
+  context->getProperty(NullValue.getName(), null_value_);
+}
+
+bool AttributesToJSON::isCoreAttributeToBeFiltered(const std::string& attribute) const {
+  return !include_core_attributes_ && core_attributes_.find(attribute) != core_attributes_.end();
+}
+
+std::unordered_set<std::string> AttributesToJSON::getAttributesToBeWritten(const std::map<std::string, std::string>& flowfile_attributes) const {
+  std::unordered_set<std::string> attributes;
+
+  for (const auto& attribute : attribute_list_) {
+    if (!isCoreAttributeToBeFiltered(attribute)) {
+      attributes.insert(attribute);
+    }
+  }
+
+  if (attributes_regular_expression_) {
+    for (const auto& [key, value] : flowfile_attributes) {
+      if (!isCoreAttributeToBeFiltered(key) && std::regex_match(key, attributes_regular_expression_.value())) {
+        attributes.insert(key);
+      }
+    }
+  }
+
+  return attributes;
+}
+
+void AttributesToJSON::addAttributeToJson(rapidjson::Document& document, const std::string& key, const std::string& value) {
+  rapidjson::Value json_key(key.c_str(), document.GetAllocator());
+  rapidjson::Value json_val;
+  if (!value.empty() || !null_value_) {
+    json_val.SetString(value.c_str(), document.GetAllocator());
+  }
+  document.AddMember(json_key, json_val, document.GetAllocator());
+}
+
+std::string AttributesToJSON::buildAttributeJsonData(std::map<std::string, std::string>&& flowfile_attributes) {
+  auto root = rapidjson::Document(rapidjson::kObjectType);
+
+  if (!attribute_list_.empty() || attributes_regular_expression_) {
+    auto attributes_to_write = getAttributesToBeWritten(flowfile_attributes);
```

Comment:
we could return an optional set from `getAttributesToBeWritten` and do the "else branch" when that is `nullopt`, this way we wouldn't need to recheck the condition that makes the result "valid" outside of that method

Reply:
Could you elaborate a bit more please? I'm not sure what do you mean by rechecking the result of `getAttributesToBeWritten`. When would it be invalid?

Reply:
currently when `attribute_list_.empty() && !attributes_regular_expression_` the return value of `getAttributesToBeWritten` (which then is an empty set) is considered "invalid" in the sense that we fallback on enumerating all the attributes on the flowfile, but of course an empty set could also mean that e.g. the regex did not match any of the attributes, so we currently cannot decide on the return value of the `getAttributesToBeWritten` if we should fallback or not

if we returned an optional from `getAttributesToBeWritten` the return value alone could signal the need for fallback

Reply:
Thanks for the explanation it is clear now, updated in effc866d914fb3ccff4f70a7a0ebcea6305403fc